### PR TITLE
Pgvector tutorial

### PR DESCRIPTION
**Pull request by TJaniF**
📚

### CODE REVIEW

File: learn/airflow-pgvector.md [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+sidebar_label: "pgvector"
+description: "Learn how to integrate pgvector and Airflow."
+id: airflow-pgvector
+sidebar_custom_props: { icon: 'img/integrations/postgres.png' }
+---
+
+import CodeBlock from '@theme/CodeBlock';
+import query_book_vectors from '!!raw-loader!../code-samples/dags/airflow-pgvector/query_book_vectors.py';
+
+[Pgvector](https://github.com/pgvector/pgvector) is an open source extension for PostgreSQL databases that adds the possibility to store and query high-dimensional object embeddings. The [Pgvector Airflow provider](https://airflow.apache.org/docs/apache-airflow-providers-pgvector/stable/index.html) offers modules to easily integrate Pgvector with Airflow.
```
Comment (kentdanas) [inline learn/airflow-pgvector.md:12]:
```suggestion
[Pgvector](https://github.com/pgvector/pgvector) is an open source extension for PostgreSQL databases that adds the possibility to store and query high-dimensional object embeddings. The [pgvector Airflow provider](https://airflow.apache.org/docs/apache-airflow-providers-pgvector/stable/index.html) offers modules to easily integrate pgvector with Airflow.
```
Seems like lowercase is how postgres writes "pgvector" - worth double checking me, but regardless let's be consistent throughout.

File: learn/airflow-pgvector.md [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+      triggerer:
+        networks:
+          - airflow
+      postgres:
+        networks:
+          - airflow
+    ```
+
+
+4. To create an [Airflow connection](connections.md) to the PostgreSQL database, add the following environment variable to your `.env` file. If you are using the OpenAI API for embeddings you also need to add your OpenAI API key to `OPENAI_API_KEY`.
```
Comment (kentdanas) [inline learn/airflow-pgvector.md:104]:
```suggestion
4. To create an [Airflow connection](connections.md) to the PostgreSQL database, add the following to your `.env` file. If you are using the OpenAI API for embeddings you will need to update the `OPENAI_API_KEY` variable.
```

File: learn/airflow-pgvector.md [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+- Basic SQL. See [SQL Tutorial](https://www.w3schools.com/sql/sql_intro.asp).
+- Vector embeddings. See [Vector Embeddings](https://tembo.io/blog/pgvector-and-embedding-solutions-with-postgres/).
+- Airflow fundamentals, such as writing DAGs and defining tasks. See [Get started with Apache Airflow](get-started-with-airflow.md).
+- Airflow decorators. [Introduction to the TaskFlow API and Airflow decorators](airflow-decorators.md).
+- Airflow connections. See [Managing your Connections in Apache Airflow](connections.md).
+
+## Prerequisites
+
+- The [Astro CLI](https://docs.astronomer.io/astro/cli/get-started).
+- An OpenAI API key of at least [tier 1](https://platform.openai.com/docs/guides/rate-limits/usage-tiers) if you want to use OpenAI for vectorization. If you do not want to use OpenAI you can adapt the `create_embeddings` function at the start of the DAG to use a different vectorizer. Note that you will likely need to adjust the `MODEL_VECTOR_LENGTH` parameter in the DAG if you use a different vectorizer.
```
Comment (kentdanas) [inline learn/airflow-pgvector.md:45]:
```suggestion
- An OpenAI API key of at least [tier 1](https://platform.openai.com/docs/guides/rate-limits/usage-tiers) if you want to use OpenAI for vectorization. If you do not want to use OpenAI you can adapt the `create_embeddings` function at the start of the DAG to use a different vectorizer.
```
I don't think anybody would remember updating the param from reading about it in the prereqs, and you've already got a note about it below.

File: learn/airflow-pgvector.md [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+    - The `get_already_imported_book_ids` task queries the `Book` table to return all `book_id` values of books that were already stored with their vectors in previous DAG runs. 
+    - The `import_book_data` task uses the [`@task` decorator](airflow-decorators.md) to read the book data from the `book_data.txt` file and return it as a list of dictionaries with keys corresponding to the columns of the `Book` table.
+
+    - The `create_embeddings_book_data` task is [dynamically mapped](dynamic-tasks.md) over the list of dictionaries returned by the `import_book_data` task to parallelize vector embedding of all book descriptions that have not been added to the `Book` table in previous DAG runs. The `create_embeddings` function defines how the embeddings are computed and can be modified to use other embedding models. If all books in the list have already been added to the `Book` table all mapped task instances are skipped.
```
Comment (kentdanas) [inline learn/airflow-pgvector.md:153]:
```suggestion
    - The `create_embeddings_book_data` task is [dynamically mapped](dynamic-tasks.md) over the list of dictionaries returned by the `import_book_data` task to parallelize vector embedding of all book descriptions that have not been added to the `Book` table in previous DAG runs. The `create_embeddings` function defines how the embeddings are computed and can be modified to use other embedding models. If all books in the list have already been added to the `Book` table, then all mapped task instances are skipped.
```

**Review by kentdanas — APPROVED** [review]
🔥 
I really have no feedback - on to copy editing!

File: learn/airflow-pgvector.md (hunk 1 above)
Comment (TJaniF) [reply learn/airflow-pgvector.md:12]: Oh, yes definitely! Even in the title of their repo Readme it is lowercase, will make sure to adjust :)

**Review by TJaniF — COMMENTED** [review]
😊 Thank you @kentdanas :) 

@lzdanski and @jwitz second one of the day :) Another vector database so same embedding + querying structure, but this time with 📚

File: learn/airflow-pgvector.md [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+id: airflow-pgvector
+sidebar_custom_props: { icon: 'img/integrations/postgres.png' }
+---
+
+import CodeBlock from '@theme/CodeBlock';
+import query_book_vectors from '!!raw-loader!../code-samples/dags/airflow-pgvector/query_book_vectors.py';
+
+[Pgvector](https://github.com/pgvector/pgvector) is an open source extension for PostgreSQL databases that adds the possibility to store and query high-dimensional object embeddings. The [pgvector Airflow provider](https://airflow.apache.org/docs/apache-airflow-providers-pgvector/stable/index.html) offers modules to easily integrate pgvector with Airflow.
+
+In this tutorial you'll use Airflow to orchestrate the embedding of book descriptions with the OpenAI API, ingest the embeddings into a PostgreSQL database with pgvector installed, and query the database for books that match a user-provided mood.
```
Comment (lzdanski) [inline learn/airflow-pgvector.md:14]:
```suggestion
In this tutorial, you use Airflow to orchestrate the embedding of book descriptions with the OpenAI API, ingest the embeddings into a PostgreSQL database with pgvector installed, and query the database for books that match a user-provided mood.
```

File: learn/airflow-pgvector.md [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+- Basic SQL. See [SQL Tutorial](https://www.w3schools.com/sql/sql_intro.asp).
+- Vector embeddings. See [Vector Embeddings](https://tembo.io/blog/pgvector-and-embedding-solutions-with-postgres/).
+- Airflow fundamentals, such as writing DAGs and defining tasks. See [Get started with Apache Airflow](get-started-with-airflow.md).
+- Airflow decorators. [Introduction to the TaskFlow API and Airflow decorators](airflow-decorators.md).
+- Airflow connections. See [Managing your Connections in Apache Airflow](connections.md).
+
+## Prerequisites
+
+- The [Astro CLI](https://docs.astronomer.io/astro/cli/get-started).
+- An OpenAI API key of at least [tier 1](https://platform.openai.com/docs/guides/rate-limits/usage-tiers) if you want to use OpenAI for vectorization. If you do not want to use OpenAI you can adapt the `create_embeddings` function at the start of the DAG to use a different vectorizer.
```
Comment (lzdanski) [inline learn/airflow-pgvector.md:45]:
```suggestion
- An OpenAI API key of at least [tier 1](https://platform.openai.com/docs/guides/rate-limits/usage-tiers) if you want to use OpenAI for vectorization. If you do not want to use OpenAI, you can adapt the `create_embeddings` function at the start of the DAG to use a different vectorizer.
```

**Review by lzdanski — COMMENTED** [review]
Just some copyediting feedback! Should be good to merge after addressing without my explicit approval

File: learn/airflow-pgvector.md [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+
+3. View your book suggestion in the task logs of the `print_book_suggestion` task:
+
+    ```text
+    [2023-11-20, 10:09:43 UTC] {logging_mixin.py:154} INFO - Book suggestion for 'A philosophical book about consciousness.':
+    [2023-11-20, 10:09:43 UTC] {logging_mixin.py:154} INFO - You should read The Idea of the World by Bernardo Kastrup, published in 2019!
+    [2023-11-20, 10:09:43 UTC] {logging_mixin.py:154} INFO - Goodreads describes the book as: A rigorous case for the primacy of mind in nature, from philosophy to neuroscience, psychology and physics. [...]
+    ```
+
+## Step 5: (Optional) Fetch and read the book
```
[the remark this answers is not in the document]
Comment (TJaniF) [reply learn/airflow-pgvector.md:182]: 😊

**Review by TJaniF — COMMENTED** [review]
Thank you @lzdanski, accepted all the suggestions :)

**Pull request merged**